src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependenciesGraphProvider.cs

```diff
@@ -0,0 +1,653 @@
+    // TODO what todo with ProjectKind here? Provide several implementations or request change to provide multiple project types?
+    //      ProjectKind = ProjectSystemPackage.ProjectTypeGuidStringWithParentheses)]
+    // without specifying ProjectType our GraphProvider will be called for all project types
+    // which might cause some extra perf overhead.
+    internal class DependenciesGraphProvider : OnceInitializedOnceDisposedAsync, IGraphProvider
+    {
+        private IDependenciesGraphProjectContextProvider ProjectContextProvider { get; set; }
+        private SVsServiceProvider ServiceProvider { get; set; }
+
+        private bool _IsInitialized;
+
+        private object _initializationLock = new object();
+
+        [ImportingConstructor]
+        public DependenciesGraphProvider(IDependenciesGraphProjectContextProvider projectContextProvider,
+                                             SVsServiceProvider serviceProvider)
+            : base(new JoinableTaskContextNode(ThreadHelper.JoinableTaskContext))
+        {
+            ProjectContextProvider = projectContextProvider;
+            ServiceProvider = serviceProvider;
+        }
+
+        /// <summary>
+        /// Remembers expanded graph nodes.
+        /// TODO: Check how to clean them when project is unloaded. (What graph is doing in general when project is killed?)
+        /// </summary>
+        private WeakCollection<IGraphContext> ExpandedGraphContexts { get; set; }
+        private HashSet<string> RegisteredSubTreeProviders { get; set; }
+
+        private object _RegisteredSubTreeProvidersLock = new object();
+
+        private object _ExpandedGraphContextsLock = new object();
+
+        /// <summary>
+        /// All icons that are used tree graph, register their monikers once and refer to them by string id
+        /// </summary>
+        private HashSet<ImageMoniker> Images { get; set; }
+
+        protected override Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            ExpandedGraphContexts = new WeakCollection<IGraphContext>();
+            RegisteredSubTreeProviders = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            Images = new HashSet<ImageMoniker>();
+
+            ProjectContextProvider.ProjectContextChanged += ProjectContextProvider_ProjectContextChanged;
+
+            return TplExtensions.CompletedTask;
+        }
+
+        protected override Task DisposeCoreAsync(bool initialized)
+        {
+            ProjectContextProvider.ProjectContextChanged -= ProjectContextProvider_ProjectContextChanged;
+
+            return TplExtensions.CompletedTask;
+        }
+
+        private async Task EnsureInitializedAsync()
+        {
+            lock (_initializationLock)
+            {
+                if (_IsInitialized)
+                {
+                    return;
+                }
+
+                _IsInitialized = true;
```

Comment:
right, changing

---
In reply to: [71808287](https://github.com/dotnet/roslyn-project-system/pull/309#discussion_r71808287) [](ancestors = 71808287)